logic/tool/TypeHinter.java

```diff
@@ -177,15 +153,15 @@ private void removeHintsViolatingSuper(Variable subVar) {
         if (supVar == null) return;
         if (supVar.reference().isLabel()) return;
 
-        Set<Label> supLabels = getHintLabels(supVar);
-        Set<Label> subLabels = getHintLabels(subVar);
-        if (supLabels == null || supLabels.isEmpty()) return;
+        Set<Label> supLabels = supVar.typeHints();
+        Set<Label> subLabels = subVar.typeHints();
+        if (supLabels.isEmpty()) return;
 
-        if (subLabels == null) return;
         if (subLabels.isEmpty()) {
             Set<Label> subHintsOfSupLabels = supLabels.stream().flatMap(label -> getType(label).getSubtypes())
                     .map(Type::getLabel).collect(Collectors.toSet());
-            addHintLabels(subVar, subHintsOfSupLabels);
+            subVar.addHints(subHintsOfSupLabels);
+            if (subVar.typeHints().isEmpty()) subVar.setSatisfiability(false);
```

Comment:
how's it possible to add hints and its set to `false`??

Reply:
`subHintsOfSupLabels` could be empty.

It's possible I might need to rename `addHintLabels` to `setHintLabels`

Reply:
On second thoughts, we've already checked if supLabels is empty by this point, so I will remove this line.